renku/service/controllers/mixins.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku service controller mixin."""
+from renku.core.utils.contexts import chdir
+from renku.service.controllers.remote_project import RemoteProject
+
+
+class ReadOperationMixin:
+    """Read operation mixin."""
+
+    def __init__(self, cache, user_data, request_data):
+        """Read operation mixin for controllers."""
+        self.user = cache.ensure_user(user_data)
+
+        self.cache = cache
+        self.user_data = user_data
+        self.request_data = request_data
+
+    @property
+    def context(self):
```

Comment:
we could add 
```
    def context(self):
        """Controller operation context."""
        return self.ctx
```
and mark the class as abstract

Reply:
```
    def to_response(self):
        """Execute controller flow and serialize to service response."""
        self.ctx["datasets"] = []

        if "project_id" in self.ctx:
            self.ctx["datasets"] = self.local()

        elif "git_url" in self.ctx:
            self.ctx["datasets"] = self.remote()

        return result_response(DatasetsListCtrl.RESPONSE_SERIALIZER, self.ctx)
```

Could also be done in the abstract class with `"datasets"` and `DatasetsListCtrl.RESPONSE_SERIALIZER` set as class attribute on the children

Reply:
I'm not really following what exactly you are suggesting here. 

Context property is already defined on `ReadOperationMixin` which serves as a partial trait for re-usability purposes, here is the implementation only part. The reason why it's here and not in the parent is because it would break the static typing constrain opposed here. We could move it there, and due to the dynamic nature of Python things would work, but my static analyzer checker would then be completely broken (same as mypy compilation). `to_response` part of the interface is meant to be implemented only on the implementation of the controller cause it contains response specific elements, so grouping it together under some abstract class would not really work here.

Reply:
Ok, I did a little restructuring for a better clarity. Under controllers modules you can find 2 additional submodules `api` and `utils`. Inside of the `api` module you can find the definition of the controller interface and a mixin traits, usually each controller will implement the controller flow (can be found inside `api.py`) and a mixin depending on an operation.

Reply:
What I mean was that all the Ctrl's are very similar, so you could have something like 
```
class ReadOperationMixin(ABC):
    def __init__(self, request_data, user_data):
        self.ctx = self.REQUEST_SERIALIZER.load(request_data)

    @abstractmethod
    def renku_op(self):
        raise NotImplementedException()

    @property
    def context(self):
        """Controller operation context."""
        return self.ctx

    def to_response(self):
        """Execute controller flow and serialize to service response."""
        self.ctx[self.CONTEXT_FIELD] = []

        if "project_id" in self.ctx:
            self.ctx[self.CONTEXT_FIELD] = self.local()

        elif "git_url" in self.ctx:
            self.ctx[self.CONTEXT_FIELD] = self.remote()

        return result_response(self.RESPONSE_SERIALIZER, self.ctx)

class DatasetsListCtrl(ReadOperationMixin):
    """Controller for datasets list endpoint."""

    REQUEST_SERIALIZER = DatasetListRequest()
    RESPONSE_SERIALIZER = DatasetListResponseRPC()
    CONTEXT_FIELD = "datasets"

    def renku_op(self):
        [...whatever command this calls and sets context with...]
```

which would make the different Ctrls more slim and reduce code duplication.

I'll have a look at the new structure :smiley:

Reply:
Yea, that's exactly what you have here (except the context field, not sure why you have that there - response serialization can be different from ctrl to ctrl).

Reply:
Well, that's very similar what you have there - with the exception of the context cause now your mixin is not statically safe.